File: common/src/main/java/au/com/codeka/warworlds/common/sim/ColonyHelper.java
package au.com.codeka.warworlds.common.sim;

import au.com.codeka.warworlds.common.proto.BuildRequest;
import au.com.codeka.warworlds.common.proto.Building;
import au.com.codeka.warworlds.common.proto.Colony;
import au.com.codeka.warworlds.common.proto.Design;
import au.com.codeka.warworlds.common.proto.Planet;

/**
 * Helper for accessing information about a {@link Colony}.
 */
public class ColonyHelper {
  public static int getMaxPopulation(Planet planet) {
    if (planet.colony == null) {
      return 0;
    }

    int population = planet.population_congeniality;

    if (planet.colony.buildings != null) {
      for (Building building : planet.colony.buildings) {
        Design design = DesignHelper.getDesign(building.design_type);
        for (Design.Effect effect : design.effect) {
          if (effect.type == Design.EffectType.POPULATION_BOOST) {
            float extraPopulation = Math.min(effect.minimum, population * effect.bonus);
            population += extraPopulation;
          }
        }
      }
    }

    return population;
  }

  public static float getDeltaMineralsPerHour(Colony colony, long now) {
    float delta = colony.delta_minerals;
    for (BuildRequest br : colony.build_requests) {
      if (br.start_time < now && br.end_time > now) {
        delta += br.delta_minerals_per_hour;
      }
    }
    return delta;
  }
}

File: common/src/main/java/au/com/codeka/warworlds/common/sim/ColonyHelper.kt
package au.com.codeka.warworlds.common.sim

import au.com.codeka.warworlds.common.proto.Colony
import au.com.codeka.warworlds.common.proto.Design
import au.com.codeka.warworlds.common.proto.Planet

/**
 * Helper for accessing information about a [Colony].
 */
object ColonyHelper {
  fun getMaxPopulation(planet: Planet): Int {
    if (planet.colony == null) {
      return 0
    }
    var population: Int = planet.population_congeniality
    if (planet.colony.buildings != null) {
      for (building in planet.colony.buildings) {
        val design = DesignHelper.getDesign(building.design_type)
        for (effect in design.effect) {
          if (effect.type === Design.EffectType.POPULATION_BOOST) {
            val extraPopulation = effect.minimum.toFloat().coerceAtMost(population * effect.bonus)
            population += extraPopulation.toInt()
          }
        }
      }
    }
    return population
  }

  fun getDeltaMineralsPerHour(colony: Colony, now: Long): Float {
    var delta: Float = colony.delta_minerals
    for (br in colony.build_requests) {
      if (br.start_time < now && br.end_time > now) {
        delta += br.delta_minerals_per_hour
      }
    }
    return delta
  }
}